### PR TITLE
[#157387718] WIP: service blacklist

### PR DESCRIPTION
**Pull request by gunzip**
## Abstract

- adds a new property "blocked_channels" to the profile model 
`{ serviceId: [ array of blocked channels ]}`

- adds logic to skip blocked channel into created_message_queue_handler

profile model: 

```
{
    "blocked_channels": {
        "5a25abf4fcc89605c082f042c49a": [
            "EMAIL",
            "WEBHOOK"
        ]
    },
    "email": "xxxxxxxxx@xxxxxxx.com",
    "is_inbox_enabled": true,
    "is_webhook_enabled": true,
    "preferred_languages": [
        "it_IT"
    ],
    "version": 46
}

```

There at least are a couple of possible implementation. The one in "Scenario 1" is implemented in the current PR.

### Scenario 1

We block notifications in the `created_message_queue_handler`, before dispatching the messages to the channel queue handlers.

**Pros**: efficient as it avoids to send notifications to the channels queue handlers
**Cons**: there are some chances that the "blocked" messages are dispatched to the user in case the act of block a channel occurs *after* the message is already dispatched to the specific channel queue handler: these handlers will send the messages without any further check.

### Scenario 2

We dispatch the notifications to the specific channel queue handlers (anyway), only then we check if the user has blocked the specific channel.

**Pros**: no chance the user wil receive any notification after blocking any channel
**Cons**: require an extra call to retrieve the user's profile in the specific channel queue handler, pollute channel queues with notifications that won't be sent

## Decisions to take

What to return in the notification-status for blocked messages ? (actually we do not save any status, which is the same behavior we get when a notification is not sent for any other reason).

## TODO

- add tests

### CODE REVIEW

**Comment by digitalcitizenship** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Remember to fix the PR title by removing WIP wording when ready
  </td>
    </tr>
  </tbody>
</table>



## Affected stories

  * 🌟 [#157387718](https://www.pivotaltracker.com/story/show/157387718): Come CIT voglio poter decidere di non ricevere notifiche da uno specifico servizio di un SAS

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by cloudify** [comment]
I think Scenario 1 is a reasonable approach.

About what we return, I'd say we keep it as is. The sender shouldn't probably know anything about what happens after he sends the message.

File: lib/created_message_queue_handler.ts [hunk 1]
```diff
@@ -247,15 +261,36 @@ export async function handleMessage(
...
+    JSON.stringify(blockedChannels)
   );
 
   //
   //  Inbox storage
   //
 
+  // check if the user has blocked inbox message storage from this service
+  const userHasBlockedMessageStorage = blockedChannels.has(
+    BlockedChannelEnum.INBOX
```
Comment (cloudify) [inline lib/created_message_queue_handler.ts]: how is this option exposed to the user?

File: lib/created_message_queue_handler.ts [hunk 2]
```diff
@@ -247,15 +261,36 @@ export async function handleMessage(
...
+    newMessageWithContent.fiscalCode,
+    JSON.stringify(blockedChannels)
   );
 
   //
   //  Inbox storage
   //
 
+  // check if the user has blocked inbox message storage from this service
+  const userHasBlockedMessageStorage = blockedChannels.has(
```
Comment (cloudify) [inline lib/created_message_queue_handler.ts]: just to be clear (while reading the code I forgot this is for a specific service) `userHasBlockedMessageStorageForService`

File: lib/created_message_queue_handler.ts (hunk 1 above)
Comment (cloudify) [reply lib/created_message_queue_handler.ts]: we need to clarify this from a UX point of view: does it makes sense to disable WEBHOOK and not INBOX? shouldn't they go together?

File: lib/created_message_queue_handler.ts [hunk 3]
```diff
@@ -276,6 +311,21 @@ export async function handleMessage(
   //  Email notification
   //
 
+  // check if the user has blocked emails sent from this service
+  // 'some(true)' in case we must send the notification by email
+  // 'none' in case the user has blocked the email channel
+  const maybeIsUnblockedEmail = fromPredicate(
```
Comment (cloudify) [inline lib/created_message_queue_handler.ts]: negative booleans are hard to keep track of (mentally): `maybeIsUnblockedEmail` -> `maybeIsEmailAllowed`

File: lib/created_message_queue_handler.ts [hunk 4]
```diff
@@ -276,6 +311,21 @@ export async function handleMessage(
...
+    (hasBlockedEmail: boolean) => !hasBlockedEmail
+  )(blockedChannels.has(BlockedChannelEnum.EMAIL));
+
+  if (isNone(maybeIsUnblockedEmail)) {
```
Comment (cloudify) [inline lib/created_message_queue_handler.ts]: try to avoid logic that is there only for logging purposes, perhaps you can move the logging to a code branch that gets executed only when `maybeIsUnblockedEmail` is empty

File: lib/created_message_queue_handler.ts [hunk 5]
```diff
@@ -291,22 +341,40 @@ export async function handleMessage(
     );
   }
 
+  const maybeUnblockedEmailNotification = maybeIsUnblockedEmail.chain(
```
Comment (cloudify) [inline lib/created_message_queue_handler.ts]: +comment since it's hard to follow what's going on

File: lib/created_message_queue_handler.ts [hunk 6]
```diff
@@ -291,22 +341,40 @@ export async function handleMessage(
...
+    _ => maybeEmailNotification
+  );
+
   //
   //  Webhook notification
   //
 
+  // check if the user has blocked webhook notifications sent from this service
+  const isBlockedWebhook = blockedChannels.has(BlockedChannelEnum.WEBHOOK);
```
Comment (cloudify) [inline lib/created_message_queue_handler.ts]: think of boolean identifiers as questions: is webhook blocked? -> `isWebhookBlocked` (also for consistency with `isWebhookEnabled`)

File: lib/created_message_queue_handler.ts [hunk 7]
```diff
@@ -276,6 +311,21 @@ export async function handleMessage(
...
 
+  // check if the user has blocked emails sent from this service
+  // 'some(true)' in case we must send the notification by email
+  // 'none' in case the user has blocked the email channel
+  const maybeIsUnblockedEmail = fromPredicate(
+    (hasBlockedEmail: boolean) => !hasBlockedEmail
+  )(blockedChannels.has(BlockedChannelEnum.EMAIL));
+
+  if (isNone(maybeIsUnblockedEmail)) {
+    winston.warn(
```
Comment (cloudify) [inline lib/created_message_queue_handler.ts]: is it something we need to be concerned about? this is probably `debug` level

File: lib/created_message_queue_handler.ts [hunk 8]
```diff
@@ -291,22 +341,40 @@ export async function handleMessage(
...
+  );
+
   //
   //  Webhook notification
   //
 
+  // check if the user has blocked webhook notifications sent from this service
+  const isBlockedWebhook = blockedChannels.has(BlockedChannelEnum.WEBHOOK);
+
+  if (isBlockedWebhook) {
```
Comment (cloudify) [inline lib/created_message_queue_handler.ts]: similar comment as above

File: lib/created_message_queue_handler.ts [hunk 9]
```diff
@@ -291,22 +341,40 @@ export async function handleMessage(
...
   // whether the recipient wants us to send notifications to the app backend
-  const isWebhookEnabled = maybeProfile.exists(
-    profile => profile.isWebhookEnabled === true
-  );
+  const isWebhookEnabled =
+    !isBlockedWebhook &&
+    maybeProfile.exists(profile => profile.isWebhookEnabled === true);
 
-  const maybeWebhookNotification = isWebhookEnabled
+  const maybeUnblockedWebhookNotification = isWebhookEnabled
```
Comment (cloudify) [inline lib/created_message_queue_handler.ts]: Unblocked/NotBlocked -> Allowed (prefer "positive" identifiers)

File: lib/models/profile.ts [hunk 10]
```diff
@@ -15,13 +15,26 @@ import { Option } from "fp-ts/lib/Option";
...
 
+const ProfileBlockedChannels = t.dictionary(
+  ServiceId,
+  readonlySetType(BlockedChannel, "ProfileBlockedChannel"),
+  "ProfileBlockedChannels"
+);
+
+// typescript does not allow to have ServiceId as key type here
+export interface IProfileBlockedChannels {
+  readonly [serviceId: string]: ReadonlySet<BlockedChannel>;
```
Comment (cloudify) [inline lib/models/profile.ts]: why not `ServiceId` instead of `string`?

File: lib/created_message_queue_handler.ts (hunk 1 above)
Comment (gunzip) [reply lib/created_message_queue_handler.ts]: another option is to replce "INBOX" with "ALL" and block all notification channels *and* the inbox.
in other words, if the inbox (message storage) is blocked, then every channel is blocked. 

cc @matteodesanti

File: lib/models/profile.ts (hunk 10 above)
Comment (gunzip) [reply lib/models/profile.ts]: because typescript does not allow to have ServiceId as key type (only strings, as stated by the comment above : )

File: lib/created_message_queue_handler.ts (hunk 1 above)
Comment (cloudify) [reply lib/created_message_queue_handler.ts]: yes, conceptually the inbox is not a notification channel (the inbox is orthogonal to the notifications) - what do you think about something like:

```ts
type BlockedInboxOrChannels = { [k: Channel]: true } | "INBOX";
```
`INBOX` implies inbox and all channels are blocked

Comment (gunzip) [reply lib/created_message_queue_handler.ts]: nice but note that openapi 2.0 spec (the version we use) does not support "oneOf" in the schema and so doesn't our typescript generator (italia-utils). it means that

1. we cannot have that type generated from the swagger specs
2. actually there's no way to enforce this in clients

Comment (cloudify) [reply lib/created_message_queue_handler.ts]: but that's an opportunity to add a new feature to our tool! 😸

Comment (gunzip) [reply lib/created_message_queue_handler.ts]: as a workaround we can just settle to union("inbox", "webhook", "email") and handle the logic server side (more or less like the actual logic but implying "webhook", "email" in case inbox is set)

Comment (cloudify) [reply lib/created_message_queue_handler.ts]: yes, less elegant (the type doesn't fully describe the logic) but I guess that's the only choice

File: lib/created_message_queue_handler.ts (hunk 4 above)
Comment (gunzip) [reply lib/created_message_queue_handler.ts]: it's used afterwards

**Comment by gunzip** [comment]
> About what we return, I'd say we keep it as is. The sender shouldn't probably know anything about what happens after he sends the message.

That's right but we may need to answer the question "why that notification we sent in the past wasn't delivered ?"

As the user may unblock services, the information that the notification wasn't delivered due to the user's past blacklist gets lost. That's why, I guess, we should save the NotificationStatus (ie. "blocked") anyway.

@cloudify

**Comment by cloudify** [comment]
> That's right but we may need to answer the question "why that notification we sent in the past wasn't delivered ?"

why we need to answer that question? notification may not be delivered for a multitude of reasons out of our control (e.g. email flagged as spam)

File: lib/created_message_queue_handler.ts [hunk 11]
```diff
@@ -17,6 +17,8 @@ import { configureAzureContextTransport } from "./utils/logging";
 
 import * as documentDbUtils from "./utils/documentdb";
 
+import { Set } from "json-set-map";
```
Comment (cloudify) [inline lib/created_message_queue_handler.ts]: We have [this io-ts set](https://teamdigitale.github.io/italia-ts-commons/modules/_types_.html#readonlysettype)

File: lib/created_message_queue_handler.ts [hunk 12]
```diff
@@ -219,6 +226,7 @@ async function createNotification(
  *          a TransientError in case of recoverable errors
  *          a PermanentError in case of unrecoverable error
  */
+// tslint:disable-next-line:cognitive-complexity
```
Comment (cloudify) [inline lib/created_message_queue_handler.ts]: 🆘

File: lib/created_message_queue_handler.ts [hunk 13]
```diff
@@ -295,23 +344,41 @@ export async function handleMessage(
...
+    );
+  }
+
   // whether the recipient wants us to send notifications to the app backend
-  const isWebhookEnabled = maybeProfile.exists(
-    profile => profile.isWebhookEnabled === true
-  );
+  const isWebhookEnabled =
+    !isWebhookBlockedForService &&
+    maybeProfile.exists(profile => profile.isWebhookEnabled === true);
```
Comment (cloudify) [inline lib/created_message_queue_handler.ts]: just for added readability, I would assign this value first to a const, e.g.:

```ts
const isWebhookEnabledInProfile = maybeProfile.exists(profile => profile.isWebhookEnabled === true);
const isWebhookEnabled = isWebhookEnabledInProfile && !isWebhookBlockedForService;
```

File: lib/created_message_queue_handler.ts (hunk 11 above)
Comment (gunzip) [reply lib/created_message_queue_handler.ts]: yes, but that's  not what we use in other models (ie. service with authorized_cidrs). we may consider a broader refactor using io-ts sets instead of this one (but we must check that io-ts sets are serializable first)

File: lib/created_message_queue_handler.ts (hunk 12 above)
Comment (gunzip) [reply lib/created_message_queue_handler.ts]: fixed. note that this come from the fact that every block ```if(...) { winston.debug }``` increases the cognitive-complexity : )
